File: killrvideo-service-comments/src/main/java/com/killrvideo/service/comment/dto/CommentByVideo.java
package com.killrvideo.service.comment.dto;

import java.util.UUID;

import com.datastax.oss.driver.api.mapper.annotations.CqlName;
import com.datastax.oss.driver.api.mapper.annotations.Entity;
import com.datastax.oss.driver.api.mapper.annotations.PartitionKey;
import lombok.NoArgsConstructor;

/**
 * Specialization for VIDEO.
 *
 * @author DataStax Developer Advocates team.
 */
@Entity
@CqlName("comments_by_video")
@NoArgsConstructor
public class CommentByVideo extends Comment {
    private static final long serialVersionUID = -6738790629520080307L;

    public static CommentByVideo from(Comment comment) {
        return new CommentByVideo(comment);
    }
    
    private CommentByVideo(Comment c) {
        super(c);
    }

    @PartitionKey
    public UUID getVideoid() {
        return videoid;
    }

    public Comment toComment() {
        return new Comment(this);
    }
}

File: killrvideo-service-comments/src/main/kotlin/com/killrvideo/service/comment/dto/CommentByVideo.kt
package com.killrvideo.service.comment.dto

import com.datastax.oss.driver.api.mapper.annotations.*
import java.time.Instant
import java.util.*
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

/**
 * Specialization for VIDEO.
 *
 * @author DataStax Developer Advocates team.
 */
@Entity
@CqlName("comments_by_video")
data class CommentByVideo(
    @PartitionKey
    var videoid: @NotNull UUID?,
    @ClusteringColumn
    var commentid: @NotNull UUID?,
    var comment: @Size(min = 1, message = "The comment must not be empty") String?,
    var userid: @NotNull UUID?,
    @Computed("toTimestamp(commentid)")
    var dateOfComment: @NotNull Instant?
) {
    fun toComment(): Comment =
        Comment(
            userid = this.userid,
            videoid = this.videoid,
            commentid = this.commentid,
            comment = this.comment,
            dateOfComment = this.dateOfComment
        )

    companion object {
        fun from(comment: Comment): CommentByVideo =
            CommentByVideo(
                userid = comment.userid,
                videoid = comment.videoid,
                commentid = comment.commentid,
                comment = comment.comment,
                dateOfComment = comment.dateOfComment
            )
    }
}
